extensions/amp-sticky-ad/0.1/amp-sticky-ad.js

```diff
@@ -37,6 +38,82 @@ class AmpStickyAd extends AMP.BaseElement {
       dev.warn(TAG, `TAG ${TAG} disabled`);
       return;
     }
+
+    // Check `amp-sticky-ad` only have one child with tagName `amp-ad`.
+    this.getRealChildren_ = this.getRealChildren();
+    if (this.getRealChildren_.length != 1 ||
+        this.getRealChildren_[0].tagName != 'AMP-AD') {
+      return;
+    }
+
+    /** @private @const {!Viewport} */
+    this.viewport_ = this.getViewport();
+
+    /** @private {boolean} */
+    this.isDisplayed_ = false;
+
+    /** @private {number} */
+    this.initialScrollTop_ = null;
+
+    /** @const @private {!Vsync} */
+    this.vsync_ = this.getVsync();
+
+    // On viewport scroll, check requirements for amp-stick-ad to display.
+    this.viewport_.onScroll(() => this.displayAfterScroll());
+  }
+
+  /** @override */
+  layoutCallback() {
+    this.isExperimentOn_ = isExperimentOn(this.getWin(), TAG);
+    if (!this.isExperimentOn_) {
+      dev.warn(TAG, `TAG ${TAG} disabled`);
+      return Promise.resolve();
+    }
+    this.scheduleLayout(this.getRealChildren_);
+    return Promise.resolve();
+  }
+
+  // The sticky ad is shown when user scroll at least one viewport and
+  // there is at least one more viewport available.
+  displayAfterScroll() {
+    if (!this.isDisplayed_) {
+      this.scrollTop_ = this.viewport_.getScrollTop();
+      this.scrollHeight_ = this.viewport_.getScrollHeight();
+      this.viewportHeight_ = this.viewport_.getSize().height;
+      if (!this.initialScrollTop_) {
+        this.initialScrollTop_ = this.scrollTop_;
```

Comment:
Hmm. This, however, is not what it does right now. There are no checks on whether user is leaving or not leaving. The scroll events are arriving continuously and we don't even know whether it's part of the same scroll sequence or multiple disjointed attempts to scroll. I'd definitely recommend simplyfing this code dramatically and to get a second stub at it in the follow-up PR.